### ISSUE TITLE
Module applications in where clauses of erased definitions yield non-erased code

### ISSUE THREAD

**Issue by nad** [agda/agda#5410]
Module applications in where clauses in erased definitions can use erased values, but the resulting definitions are not treated as erased. The following code should not be accepted:
```agda
module _ where

open import Agda.Builtin.Bool

module M (b : Bool) where

  some-boolean : Bool
  some-boolean = b

@0 A : @0 Bool → Set
A b = Bool
  module A where
    module M′ = M b

bad : @0 Bool → Bool
bad = A.M′.some-boolean
```

Related: #4743.

**Comment by nad** [comment]
I tried to fix this by making the definitions arising from a module application erased if the module application was made in an erased setting. However, this does not work very well. With my change in place the following code was rejected:
```agda
{-# OPTIONS --without-K #-}

module _ where

module M where
  data D : Set where

record R : Set where
  field
    x : let module M′ = M in M′.D
```
This is similar to the problems with pattern-matching lambdas that we have discussed. We solved this by introducing erased pattern-matching lambdas, and I suppose we could do the same for modules. There is already an issue for that (#4743).

**Comment by nad** [comment]
I tried another fix, in which module applications are always type-checked in non-erased contexts. This made the following code fail:
```agda
open import Agda.Primitive

variable
  ℓ : Level

record R ℓ : Set ℓ where

A : (r : R ℓ) → let open R r in Set₁
A _ = Set
```
```
Failed to solve the following constraints:
  _ℓ_5 = _ℓ_8 : Level (blocked on any(_ℓ_5, _ℓ_8))
when checking the module application module _ = R r
```
The output produced by Agda when `-vtc:100` is used includes the following lines:
```
occursCheck: meta variable _ℓ_5 has relevance and quantity @0
Meta occurs check found bad relevance
aborting because meta is frozen
```
@jespercockx, you recently worked on the code that produces that output. Do you have any ideas about how this could be fixed?

The problem only seems to affect one definition in the standard library, so an alternative might be to accept the regression and change the type signature:
```agda
A : ∀ {ℓ} (r : R ℓ) → let open R r in Set₁
A _ = Set
```

**Comment by jespercockx** [comment]
I don't understand the details 100%, but since the debug message mentions that the meta is frozen I would guess that the actual problem occurs earlier in the process. What is the output of `-vtc.meta.occurs:50`?

**Comment by nad** [comment]
I switched to a test case that does not make explicit use of `Level`s:
```agda
variable
  A : Set

record R (A : Set) : Set where

_ : (r : R A) → let open R r in Set₁
_ = λ _ → Set
```
Here is the output produced by Agda with `-vtc.meta.occurs:50`:
```
Checking Issue5410-2 ([…]/Issue5410-2.agda).
occursCheck MetaId 0 VarMap {theVarMap = fromList []}
initOccursCheck: we look into the following definitions:
(none)
occursCheck _0 (StronglyRigid) of  Issue5410-2.GeneralizeTel
occursCheck MetaId 3 VarMap {theVarMap = fromList [(0,VarOcc {varFlexRig = Unguarded, varModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}})]}
initOccursCheck: we look into the following definitions:

occursCheck _3 (StronglyRigid) of  Set
occursCheck MetaId 5 VarMap {theVarMap = fromList [(0,VarOcc {varFlexRig = Unguarded, varModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}}),(1,VarOcc {varFlexRig = Unguarded, varModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}})]}
initOccursCheck: we look into the following definitions:

occursCheck _5 (StronglyRigid) of  _A_2
occursCheck: meta variable _A_2 has relevance and quantity @0
Meta occurs check found bad relevance
aborting because meta is frozen
initOccursCheck: we look into the following definitions:

occursCheck _5 (StronglyRigid) of  _A_2
occursCheck: meta variable _A_2 has relevance and quantity @0
Meta occurs check found bad relevance
aborting because meta is frozen
occursCheck MetaId 5 VarMap {theVarMap = fromList [(0,VarOcc {varFlexRig = Unguarded, varModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}}),(1,VarOcc {varFlexRig = Unguarded, varModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}})]}
initOccursCheck: we look into the following definitions:

occursCheck _5 (StronglyRigid) of  _A_2
occursCheck: meta variable _A_2 has relevance and quantity @0
Meta occurs check found bad relevance
aborting because meta is frozen
initOccursCheck: we look into the following definitions:

occursCheck _5 (StronglyRigid) of  _A_2
occursCheck: meta variable _A_2 has relevance and quantity @0
Meta occurs check found bad relevance
aborting because meta is frozen
[…]/Issue5410-2.agda:6,21-29
Failed to solve the following constraints:
  _A_2 =< _A_5 : Set (blocked on any(_A_2, _A_5))
when checking the module application module _ = R r
```

**Comment by nad** [comment]
Meta-variable 2 is frozen by the generalisation machinery right after it is created.

**Comment by jespercockx** [comment]
Here's what I think is going on:

- The meta-variable A_2 comes from the generalized variable `A`, it is marked as erased because it is created at the type level
- The meta-variable A_5 comes from the implicit argument of the module application `R r`, normally it is also marked as erased but with your change it isn't.
- We cannot instantiate A_2 := A_5 because A_5 has an extra variable in scope (`r`)
- We cannot instantiate A_5 := A_2 directly because A_2 is marked as erased, instead `A_2` needs to be promoted to non-erased first.
- However, named generalized metas are created frozen (see https://github.com/agda/agda/blob/465e46a20b59160ea38bd50a30beaab849d176c5/src/full/Agda/TypeChecking/Generalize.hs#L716-L717) so the promotion fails.

The thing that I'm still confused about is why `A_2` is marked as erased. I did a recent change to make generalize variables use their declared modality, not the modality of the context where they appear (see https://github.com/agda/agda/commit/b87190c5c7e9893ec807a272876a959eaedc9950#diff-339a2e4078ae865916f6fdd7424e241ccc12f87246f2ae663e68574976094eeb). So in this case it seems it should just get the default modality, not the erased one. So the next step would be to look at the output of `-vtc.generalize:50` and see why `A_2` was actually marked as erased.

**Comment by nad** [comment]
The following call uses the quantity ω:

https://github.com/agda/agda/blob/465e46a20b59160ea38bd50a30beaab849d176c5/src/full/Agda/TypeChecking/Generalize.hs#L728-L730

However, that call sets the quantity that will be used if the meta-variable is turned into an argument later. In this case that does not happen. The meta-variable's quantity is presumably set by `newNamedValueMeta` above:

https://github.com/agda/agda/blob/465e46a20b59160ea38bd50a30beaab849d176c5/src/full/Agda/TypeChecking/Generalize.hs#L714

This call takes the quantity from the environment. Are you suggesting that the meta-variable's quantity should be the declared one, not the one taken from the environment?

**Comment by jespercockx** [comment]
Yes, I think that would make sense. Though perhaps to be safe there should be a check that the modality of the variable is usable at the location where the meta is created, otherwise you could run into trouble for e.g. variables that are declared as irrelevant.

**Comment by nad** [comment]
Are you sure that such a check is necessary? Can you suggest some code that would be incorrectly accepted in the absence of such a check?

When I added such a check the following code was rejected:
```agda
open import Agda.Primitive

variable
  @0 a : Level
  A    : Set a
```
This can be fixed by moving `workOnTypes` to the outside of the following expression, but I don't know if that creates any other problem:

https://github.com/agda/agda/blob/465e46a20b59160ea38bd50a30beaab849d176c5/src/full/Agda/TypeChecking/Rules/Decl.hs#L542-L543

Another problem with the check is that the following code is rejected, even though `A` is only used in an erased position:
```agda
variable
  @0 A : Set

data D : @0 Set → Set₁ where
  d : D A
```
One could perhaps keep track of which modalities are in effect at the generalisable variables' use sites. However, I'm not sure exactly what the check is intended to accomplish, so I don't know if that approach is reasonable.

**Comment by jespercockx** [comment]
I was thinking of an example like this, which would perhaps be accepted if we're not careful:
```agda
postulate
  A : Set

variable
  .x : A

data This {A : Set} : A → Set where
  this : (x : A) → This x

f : This x
f = this _
```
But perhaps there are already other mechanisms in place that prevent this kind of invalid uses of variables, in which case the extra check would not be necessary.

**Comment by nad** [comment]
If type signatures with generalisable variables are re-checked after the variables have been generalised, then this kind of problem should be caught, right? I guess that such re-checking currently does not take place.

**Comment by jespercockx** [comment]
Not unless you enable double-check. But I propose we try to generate well-typed code in the first place, instead of detecting it is ill-typed after the fact.

**Comment by nad** [comment]
> I was thinking of an example like this, which would perhaps be accepted if we're not careful:

That particular piece of code is rejected without an extra check:
```
Identifier x is declared irrelevant, so it cannot be used here
when checking that the expression x has type _A_10
```
However, perhaps another, more problematic example can be constructed.

**Comment by nad** [comment]
I noticed that, with the code in my latest commit above, the following code is accepted:
```agda
module _ where

module M (A : Set) where

_ : {@0 A : Set} → let open M A in Set₁
_ = Set
```
However, the following code is rejected:
```agda
variable
  @0 A : Set

module M (A : Set) where

_ : let open M A in Set₁
_ = Set
```

**Comment by nad** [comment]
> I noticed that, with the code in my latest commit above, the following code is accepted:

This is related to issue #5341.

**Comment by nad** [comment]
> But perhaps there are already other mechanisms in place that prevent this kind of invalid uses of variables, in which case the extra check would not be necessary.

Let us consider some cases:
* A generalisable variable with the default modality. The associated meta-variable has the default modality, which I think means that there are restrictions on how it can be instantiated, but that it can be used anywhere. The resulting Π type, if any, will have a non-erased and non-irrelevant domain. I guess this is unproblematic for erasure. I don't know how irrelevance works, so I don't know if this is OK for irrelevance.
* A generalisable variable that is marked as erased. The associated meta-variable is erased, which hopefully means that it can be instantiated with erased or non-erased things, but that it can only be used in erased contexts. The resulting Π type, if any, will have an erased domain. I guess this is unproblematic.
* A generalisable variable that is marked as irrelevant. The associated meta-variable is irrelevant, and the resulting Π type, if any, will have an irrelevant domain. I don't know how irrelevance works, so I don't know if this is OK.

Given that the problem in the OP refers to erasure, and that the fix that we have discussed might work for erasure, I suggest that we implement it only for erasure.

**Comment by nad** [comment]
I noticed that constructors are treated differently. I opened a new issue related to this (#5434).
